Allow connecting to abstract namespace sockets

dennisjac opened the pull request:
Hi,
I'm currently using node-unix-dgram to implement systemd service notification support for javascript and ran into the problem that systemd can use abstract namespace sockets instead of filesystem based ones. The difference is that for abstract namespace sockets the path begins with a 0-byte but other than that they behave like regular filesystem based sockets.
I made some minor changes to the Connect function so that it works correctly in this case. These changes are transparent and shouldn't affect the use of filesystem based sockets in any way (I tested both cases).

Regards,
  Dennis

santigimeno commented on src/unix_dgram.cc:
```diff
@@ -334,20 +334,29 @@ NAN_METHOD(Connect) {
 
   memset(&s, 0, sizeof(s));
-  strncpy(s.sun_path, *path, sizeof(s.sun_path) - 1);
+  memcpy(s.sun_path, *path, len);
```
Shouldn't it be `if (len >= sizeof(s.sun_path))` to be sure s.sun_path is always null-terminated?

santigimeno replied:
Maybe including the size check to the `bind` and `send` methods is a good idea too?

dennisjac replied:
Since abstract namespace sockets don't need to be null terminated the cases probably need to be handled separately e.g. like:
```c
if (((*path)[0] == '\0' && len > sizeof(s.sun_path)) ||
    ((*path)[0] != '\0' && len >= sizeof(s.sun_path))) {
  err = -EINVAL;
  goto out;
}
```
The ```send``` method doesn't use a path did you mean the ```sendto``` method instead? Yes I could add the same handling to ```bind``` and ```sendto```.

santigimeno replied:
> Since abstract namespace sockets don't need to be null terminated the cases probably need to be handled separately 

It looks like it.

> The send method doesn't use a path did you mean the sendto method instead?

Yes

santigimeno commented on src/unix_dgram.cc:
```diff
@@ -334,20 +334,29 @@ NAN_METHOD(Connect) {
   err = 0;
-  if (connect(fd, reinterpret_cast<sockaddr*>(&s), sizeof(s)))
+  namelen = offsetof(struct sockaddr_un, sun_path) + len;
+  if (connect(fd, reinterpret_cast<sockaddr*>(&s), namelen))
```
Is there a reason not to use `sizeof(s)` as `addrlen`?

dennisjac replied:
Since abstract namespace sockets are not 0-terminated the addrlen parameter of connect() is used to determine the length. From the socket man page: "The socket's address in this namespace is given by the additional bytes in sun_path that are covered by the specified length of the address structure."

santigimeno replied:
@dennisjac I think I understand now. Thanks :). So the way `addrlen` was calculated was not strictly correct but worked because `sun_path` had to be zero-terminated, right?

dennisjac replied:
Correct :). The ```sizeof(s)``` is technically larger than necessary (at least most of the time) but practically that doesn't matter because of the 0-termination of the path.
I'm going to update the PR soon with the changes discussed so far.

dennisjac commented:
I updated the PR so that abstract namespace sockets are now fully supported for ```sendTo``` and ```bind``` as well.
I also had to modify onRecv() so that the source address is correctly returned for the abstract namespace socket case as well and I added a test.

santigimeno commented on src/unix_dgram.cc:
```diff
@@ -224,20 +229,31 @@ NAN_METHOD(Bind) {
+
   memset(&s, 0, sizeof(s));
-  strncpy(s.sun_path, *path, sizeof(s.sun_path) - 1);
+  memcpy(s.sun_path, *path, len);
```
As this is repeated multiple times, what about using an inline function?

```c++
inline int addrSetup(String::Utf8Value& path, sockaddr_un& s) {
  unsigned int len;

  len = path.length();
  if (((*path)[0] == '\0' && len > sizeof(s.sun_path)) ||
      ((*path)[0] != '\0' && len >= sizeof(s.sun_path))) {
    return -EINVAL;
  }

  memset(&s, 0, sizeof(s));
  memcpy(s.sun_path, *path, len);

  return offsetof(struct sockaddr_un, sun_path) + len;
}
```

santigimeno commented on src/unix_dgram.cc:
```diff
@@ -224,20 +229,31 @@ NAN_METHOD(Bind) {
+    err = -EINVAL;
+    goto out;
+  }
+  namelen = offsetof(struct sockaddr_un, sun_path) + len;
```
There's already an `offset_of` macro defined in the file. I don't know which should be used. @bnoordhuis ?

bnoordhuis replied:
`offsetof` should be fine.  I think I added `offset_of` to work around a bug in gcc 3.3 or 3.4's `__builtin_offsetof`.  It doesn't look like it's even used at the moment so it could be removed.

santigimeno commented:
The `test-abstract.js` should be added to the scripts -> test entry in `package.json`.
Also testing the `EINVAL` case would be nice.

dennisjac commented:
I've added the inline function and modified package.json to include the test-abstract.js file.

santigimeno commented on src/unix_dgram.cc:
```diff
@@ -219,25 +224,48 @@ NAN_METHOD(Socket) {
+    err = -EINVAL;
+    goto out;
+  }
+  namelen = res;
```
Maybe you can get rid of the `res` variable:
```c++
err = addrSetup(path, s);
if (err < 0)
  goto out;
```

santigimeno commented on src/unix_dgram.cc:
```diff
@@ -108,8 +108,13 @@ void OnRecv(SocketContext* sc) {
+      } else {
+        argv[2] = Nan::New<String>(u_addr.sun.sun_path,
+                    msg.msg_namelen-sizeof(sa_family_t)).ToLocalChecked();
+      }
```
I would add a comment here explaining that abstract sockets starts with `\0` or something like that. (If you don't feel like it may not be not that important, but I think it can be helpful for people looking at the code)

santigimeno commented on src/unix_dgram.cc:
```diff
@@ -219,25 +224,48 @@ NAN_METHOD(Socket) {
+  if (((*path)[0] == '\0' && len > sizeof(s.sun_path)) ||
+      ((*path)[0] != '\0' && len >= sizeof(s.sun_path))) {
+    return -EINVAL;
+  }
```
I would add the comment here too

santigimeno commented:
Still, a test for the `EINVAL` case would be nice.

dennisjac commented:
Removed the ```res``` variable (not sure why I didn't see that I could use ```err``` the first time) and added a test for the EINVAL case.

santigimeno commented on test/test-invalid-socketname.js:
```diff
@@ -0,0 +1,27 @@
+
+var unix = require('../lib/unix_dgram');
+var SOCKNAME = '\0abstractsocket';
+var SOCKNAME_INVALID = '\0'+(Array(120).join("a"));
```
add space surrounding `+`

dennisjac replied:
Added the spaces.

santigimeno commented:
It LGTM with the nit addressed, but I would prefer @bnoordhuis to sign off on this before merging.
Thanks!!

mvduin commented:
Since "paths" in the abstract namespace are basically just opaque binary data, wouldn't it be more appropriate to use a buffer rather than a string for them (in recv) ?

santigimeno commented:
@mvduin It makes sense. The problem I see is that the API would be different for UNIX and abstract sockets and that could be confusing.

redneb commented:
Any news on this one?

bnoordhuis commented:
I'm okay with this feature but the PR needs a rebase.
